### PR TITLE
[simulation] support virtual time UART

### PR DESCRIPTION
**Pull request by simonlingoogle**
This PR will enhance OTNS to support both virtual time UART and non-virtual time UART. 

- OTNS turns on/off virtual time UART if OS env `VIRTUAL_TIME_UART=1/0`
- Use longer `go` durations for non virtual-time UART to reduce the chance of false fails.
- Run each test for both `VIRTUAL_TIME_UART=1/0`

### CODE REVIEW

File: pylibs/unittests/test_basic.py [hunk 1]
```diff
@@ -52,10 +52,14 @@ def testGetSetMDR(self):
         assert ns.packet_loss_ratio == 1
 
     def testOneNode(self):
-        ns = self.ns
-        ns.add("router")
-        ns.go(10)
-        self.assertFormPartitions(1)
+        for i in range(1000):
```
Comment (simonlingoogle) [inline pylibs/unittests/test_basic.py:55]: Without `virtual time UART`, this test is very unlikely to pass for 1000 times.

File: script/test [hunk 2]
```diff
@@ -68,7 +68,7 @@ install() {
 get_openthread() {
   if ! [[ -x $OT_DIR ]]; then
     mkdir -p "$(dirname $OT_DIR)"
-    git clone https://github.com/openthread/openthread.git $OT_DIR
+    git clone https://github.com/simonlingoogle/openthread.git --depth 1 --branch otns/use-virtual-time-uart $OT_DIR
```
Comment (simonlingoogle) [inline script/test:71]: Can not use OT master yet because it is forbidden to compile with both `OTNS=1` and `VIRTUAL_TIME_UART=1`

**Comment by simonlingoogle** [comment]
This PR and [OT #4818](https://github.com/openthread/openthread/pull/4818) requires each other to work. There are difficulties in merging them. Which ever merge first, we leave `OTNS` master unusable for a short period of time.

On solution is that we can create an PR to make OT compile for `OTNS` with or without VIRTUAL_TIME_UART. Only after this PR is merged, we can create another OT PR to require only `VIRTUAL_TIME_UART=1` for `OTNS`. But it creates many intermediate PRs. 

@librasungirl @jwhui @wgtdkp Any suggestion on how we can better handle this kind of situation?

**Comment by wgtdkp** [comment]
@simonlingoogle I think we can schedule to merge this PR and OT#4818 at the same time.

**Comment by jwhui** [comment]
> This PR and [OT #4818](https://github.com/openthread/openthread/pull/4818) requires each other to work. There are difficulties in merging them. Which ever merge first, we leave `OTNS` master unusable for a short period of time.
> 
> On solution is that we can create an PR to make OT compile for `OTNS` with or without VIRTUAL_TIME_UART. Only after this PR is merged, we can create another OT PR to require only `VIRTUAL_TIME_UART=1` for `OTNS`. But it creates many intermediate PRs.
> 
> @librasungirl @jwhui @wgtdkp Any suggestion on how we can better handle this kind of situation?

We can:

1. Merge https://github.com/openthread/openthread/pull/4818
2. Update this PR to point to master.
3. Merge this PR.

Does that sound OK?

**Comment by simonlingoogle** [comment]
> 1. Merge [openthread/openthread#4818](https://github.com/openthread/openthread/pull/4818)
> 2. Update this PR to point to master.
> 3. Merge this PR.

Sounds good to me. 
The time period between step 1 and step 2 should better be minimized. We should do these steps when this PR is properly reviewed.

**Comment by simonlingoogle** [comment]
Hi @jwhui, @wgtdkp , 
Unlike what we previously discussed about OTNS switching to virtual time UART, this PR actually enhances OTNS to support both modes. 
The enhancement should be pretty simple (< 200 lines of code). It is now well tested and ready for reviewing. 
Please take a look when you have time, thanks.

File: dispatcher/dispatcher.go [hunk 3]
```diff
@@ -789,7 +826,22 @@ func (d *Dispatcher) handleStatusPush(srcid NodeId, data string) {
 func (d *Dispatcher) AddNode(nodeid NodeId, x, y int, radioRange int, mode NodeMode) {
 	simplelogger.AssertNil(d.nodes[nodeid])
 	simplelogger.Infof("dispatcher add node %d", nodeid)
-	d.newNode(nodeid, x, y, radioRange, mode)
+	node := d.newNode(nodeid, x, y, radioRange, mode)
+
+	// wait until node's extended address is emitted
+	// this helps OTNS to make sure that the child process is ready to receive UDP events
```
Comment (wgtdkp) [inline dispatcher/dispatcher.go:832]:
```suggestion
	// Wait until node's extended address is emitted.
	// This helps OTNS to make sure that the child process is ready to receive UDP events.
```

File: pylibs/unittests/test_basic.py [hunk 4]
```diff
@@ -24,7 +24,8 @@
 # CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 # ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 # POSSIBILITY OF SUCH DAMAGE.
-
+import logging
```
Comment (wgtdkp) [inline pylibs/unittests/test_basic.py:27]:
```suggestion
#
import logging
```

File: pylibs/unittests/test_signals.py [hunk 5]
```diff
@@ -100,10 +101,10 @@ def _test_signal_ignore(self, sig: int):
         t.join()
         self.assertRaises(TimeoutExpired, self.ns._otns.wait, timeout=0.1)
 
-        self.ns._otns.send_signal(signal.SIGKILL)
+        self.ns._otns.send_signal(signal.SIGTERM)
         t2.join()
         exit_code = self.ns._otns.wait()
-        self.assertNotEqual(0, exit_code, "exit code should not be 0")
+        self.assertEqual(0, exit_code, "exit code should not be 0")
```
Comment (wgtdkp) [inline pylibs/unittests/test_signals.py:107]:
```suggestion
        self.assertEqual(0, exit_code, "exit code should be 0")
```

File: pylibs/unittests/test_signals.py [hunk 6]
```diff
@@ -121,6 +122,8 @@ def _test_signal_exit(self, sig: int, duration: float = 1):
 
         if sig == signal.SIGKILL:
             self.assertNotEqual(0, exit_code, "exit code should not be 0")
+            # kill all ot-cli-ftd child processes
```
Comment (wgtdkp) [inline pylibs/unittests/test_signals.py:125]:
```suggestion

            # Kill all ot-cli-ftd child processes.
```

File: simulation/node.go [hunk 7]
```diff
@@ -65,17 +65,31 @@ func newNode(s *Simulation, id NodeId, cfg *NodeConfig) (*Node, error) {
 	}
 	simplelogger.Debugf("node exe path: %s", exePath)
 	cmd := exec.CommandContext(context.Background(), exePath, strconv.Itoa(id))
-	pipeIn, err := cmd.StdinPipe()
-	if err != nil {
-		return nil, err
+
+	n := &Node{
```
Comment (wgtdkp) [inline simulation/node.go:69]: What about name it `node`? `n` usually represents a number.

File: pylibs/unittests/test_basic.py [hunk 8]
```diff
@@ -51,33 +52,41 @@ def testGetSetMDR(self):
...
+            ns.go(3)
+            self.assertFormPartitions(1)
+            self.tearDown()
+            self.setUp()
 
     def testAddNode(self):
         ns = self.ns
         ns.add("router")
-        ns.go(10)
+        self.go_conservative(3)
```
Comment (wgtdkp) [inline pylibs/unittests/test_basic.py:72]:
```suggestion
        self.goConservative(3)
```

**Comment by simonlingoogle** [comment]
rebased to resolve conflicts. 
Any further comments? @wgtdkp

File: dispatcher/dispatcher.go [hunk 9]
```diff
@@ -526,6 +535,34 @@ func (d *Dispatcher) advanceNodeTime(id NodeId, timestamp uint64, force bool) {
 	}
 }
 
+// SendUART sends data to virtual time UART of the target node.
+func (d *Dispatcher) SendUART(id NodeId, data []byte) {
```
Comment (wgtdkp) [inline dispatcher/dispatcher.go:539]:
```suggestion
func (d *Dispatcher) SendToUART(id NodeId, data []byte) {
```

File: dispatcher/dispatcher.go [hunk 10]
```diff
@@ -1071,3 +1123,15 @@ func (d *Dispatcher) SetVisualizationOptions(opts VisualizationOptions) {
...
 }
+
+func (d *Dispatcher) handleUartWrite(nodeid NodeId, data []byte) {
+	if !d.cfg.VirtualTimeUART {
+		simplelogger.Fatalf("should use Virtual Time UART: -virtual-time-uart=true")
+	}
+	d.cbHandler.OnUartWrite(nodeid, data)
+}
+
+// NotifyExit notifies the dispatcher that the node process has exit.
```
Comment (wgtdkp) [inline dispatcher/dispatcher.go:1134]:
```suggestion
// NotifyExit notifies the dispatcher that the node process has exited.
```

File: otns_main/otns_main.go [hunk 11]
```diff
@@ -194,7 +194,20 @@ func createSimulation(ctx *progctx.ProgCtx) *simulation.Simulation {
...
 	simcfg.Real = args.Real
 
+	simcfg.VirtualTimeUART = parseVirtualTimeUARTFromEnviron()
+
 	sim, err := simulation.NewSimulation(ctx, simcfg)
 	simplelogger.FatalIfError(err)
 	return sim
 }
+
+func parseVirtualTimeUARTFromEnviron() bool {
```
Comment (wgtdkp) [inline otns_main/otns_main.go:204]:
```suggestion
func parseVirtualTimeUARTFromEnv() bool {
```

**Pull request merged**